pkg/server/server.go

```diff
@@ -484,29 +489,66 @@ func (p *ec2ProviderImpl) setPrivateDNSNameCache(id string, privateDNSName strin
 }
 
 // GetPrivateDNS looks up the private DNS from the EC2 API
-func (p *ec2ProviderImpl) getPrivateDNSName(id string) (string, error) {
+func (p *ec2ProviderImpl) getPrivateDNSName(id string, regions []string) (string, error) {
 	privateDNSName, err := p.getPrivateDNSNameCache(id)
 	if err == nil {
 		return privateDNSName, nil
 	}
+	var instance *ec2.Instance
 
 	// Look up instance from EC2 API
-	output, err := p.ec2.DescribeInstances(&ec2.DescribeInstancesInput{
+	if len(regions) == 0 {
+		instance, err = p.findInstanceByID(id, nil)
+		if err != nil {
+			return "", fmt.Errorf("failed querying private DNS from EC2 API for node %s: %s", id, err.Error())
+		}
+	} else {
+		for _, region := range regions {
+			instance, err = p.findInstanceByID(id, &region)
+			if err != nil {
+				return "", fmt.Errorf("failed querying private DNS from EC2 API for node %s in %s: %s", id, region, err.Error())
+			}
+			if instance != nil {
+				break
+			}
+		}
+	}
+
+	if instance == nil {
+		return "", fmt.Errorf("failed to find node %s", id)
+	}
+
+	privateDNSName = aws.StringValue(instance.PrivateDnsName)
+	p.setPrivateDNSNameCache(id, privateDNSName)
+	return privateDNSName, nil
+}
+
+func (p *ec2ProviderImpl) findInstanceByID(id string, region *string) (*ec2.Instance, error) {
+	output, err := p.getEC2Service(p.roleARN, region).DescribeInstances(&ec2.DescribeInstancesInput{
 		InstanceIds: aws.StringSlice([]string{id}),
 	})
 	if err != nil {
-		return "", fmt.Errorf("failed querying private DNS from EC2 API for node %s: %s", id, err.Error())
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "InvalidInstanceID.NotFound" {
+			// "not found" is not an error in this case
```

Comment:
Don't like this pattern of returning `nil, nil`, would prefer if it defined an error string/type and had the caller check it. `findInstanceByID` sounds to me like it should either find the instance or give me an error, not neither.